File: php/testcode/ldapi_043.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_ldapi_uuid_lookup
function ldapi043(BenchmarkRequest $req): BenchmarkResponse {
    $conn = ldap_connect('ldap://localhost');
    $dnMap = [
        'sales' => 'ou=sales,dc=example,dc=com',
        'eng' => 'ou=engineering,dc=example,dc=com',
        'hr' => 'ou=hr,dc=example,dc=com',
    ];
    $id = $req->param('id');
    $dn = $dnMap[$id] ?? null; // vuln-code-snippet safe-line php_ldapi_uuid_lookup
    if (!$dn) {
        return BenchmarkResponse::badRequest('unknown department');
    }
    $result = ldap_search($conn, $dn, '(objectClass=*)');
    $entries = ldap_get_entries($conn, $result);
    return BenchmarkResponse::json(['count' => $entries['count']]);
}
// vuln-code-snippet end php_ldapi_uuid_lookup

File: rust/testcode/ldapi_043.rs
//! CWE-90: Variable overwrite — tainted filter string replaced with hardcoded safe value before LDAP call.

fn ldap_search(base: &str, filter: &str) -> String {
    // In production: ldap3::LdapConn::new("ldap://...").search(base, Scope::Subtree, filter, vec!["cn", "mail"])
    format!("LDAP search in {} with filter {}", base, filter)
}

// vuln-code-snippet start testcodeLdapi043
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let user = req.param("username");
    let mut filter = format!("(uid={})", user);
    filter = "(objectClass=person)".to_string();
    let result = ldap_search("dc=example,dc=com", &filter); // vuln-code-snippet target-line testcodeLdapi043
    super::shared::BenchmarkResponse::ok(&result)
}
// vuln-code-snippet end testcodeLdapi043
